### PR TITLE
Override template source maps

### PR DESCRIPTION
**Pull request by zarend**
FYI: I'm putting this PR on the back-burner for now until we land [40585](https://github.com/angular/angular/pull/40585). That's because @alxhub deleted `overrideComponentTemplate` 💥 , which means much of this PR will need to be rewritten. Ty all for reviewing. I'll re-request reviews when this is ready again.

This is a 2-part PR about template source maps. The primary goal here is to cleanup our code for template sourcemaps. The secondary goal is to fix our diagnostics created when overriding a template. Commit 1 is a big commit that does a refactor. Commit 2 is the smaller commit which fixes the diagnostics override template produces and validates that the refactor is worthwhile.

## fix(language-service): don't hardcode 'override.html' for override template diagnostics

This fixes how we create diagnostics in the override template flow. It

does not affect template diagnostics created any other way.

This is the previous behavior when a template is overriden with a parse
error:
'Parser Error: Bindings cannot contain assignments at column 6 in [{{nope=false}}] in override.html@0:0'

This is the corrected error.
'Parser Error: Bindings cannot contain assignments at column 6 in [{{nope=false}}] in /app.html@0:0'

### CODE REVIEW

File: packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import {ConstantPool} from '@angular/compiler';
+import {TemplateSourceRegistryImpl} from '@angular/compiler-cli/src/ngtsc/annotations/src/template_source_registry';
```
Comment (petebacondarwin) [inline packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:9]: drive-by comment: imports from ngcc to ngtsc should be relative and target the barrel for that area

```suggestion
import {TemplateSourceRegistryImpl} from '../../../src/ngtsc/annotations';
```

File: packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+export function getLineAndCharacterFromPosition(lineStartsMap: number[], position: number) {
+  const lineIndex = findClosestLineStartPosition(lineStartsMap, position);
+  return {character: position - lineStartsMap[lineIndex], line: lineIndex};
+}
+
+/**
+ * Computes the line start map of the given text. This can be used in order to
+ * retrieve the line and character of a given text position index.
+ */
+export function computeLineStartsMap(text: string): number[] {
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:21]: There is actually a function in ngtsc/sourcemaps that does this. Perhaps we should export that from the `sourcemaps` barrel and use it here instead of redefining it?

See https://github.com/angular/angular/blob/3cfb4b4ef9127c0eca34ad2edf93004a0eaab4c5/packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file.ts#L421

File: packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  }
+  result.push(pos);
+  return result;
+}
+
+/** Finds the closest line start for the given position. */
+function findClosestLineStartPosition<T>(
+    linesMap: T[], position: T, low = 0, high = linesMap.length - 1) {
+  while (low <= high) {
+    const pivotIdx = Math.floor((low + high) / 2);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:45]:
```suggestion
    const pivotIdx = (low + high) >> 1;
```

File: packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  // In case there was no exact match, return the closest "lower" line index. We also
+  // subtract the index by one because want the index of the previous line start.
+  return low - 1;
+}
+
+/**
+ * Represents the source of a template that was processed during type-checking. This information is
+ * used when translating parse offsets in diagnostics back to their original line/column location.
+ */
+// TODO: delete me, this is duplicated from template_source_registry.ts
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:66]: This *is* the template_source_registry.ts 😸 !

File: packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  /**
+   * This map keeps track of all template sources that have been type-checked by the id that is
+   * attached to a TCB's function declaration as leading trivia. This enables translation of
+   * diagnostics produced for TCB code to their source location in the template.
+   */
+  private templateSources = new Map<ClassDeclaration, TemplateSource>();
+
+  captureSource(node: ClassDeclaration, mapping: TemplateSourceMapping, file: ParseSourceFile):
+      ClassDeclaration {
+    this.templateSources.set(node, new TemplateSource(mapping, file));
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:102]: Does this mean that a new `TemplateSource` will be generated for every class declaration in a file? Perhaps this should be cached here based on the `file`?

File: packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts (hunk 4 above)
Comment (zarend) [reply packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:66]: `// TODO: delete irrelevant comments` 😂

File: packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts (hunk 1 above)
Comment (zarend) [reply packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:9]: thanks, I fixed the imports

**Comment by zarend** [comment]
### TODO
There are the items todo before this PR will be ready to review.
 - sort out `ts.ClassDeclaration` vs the `ClassDeclaration` from reflection
 - fix TODO's in code comments
 - add unit tests for inline and indirect templates
 - squash into a single commit

**Comment by zarend** [comment]
TODO move TemplateSourceRegistry to typecheck directory

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 6]
```diff
@@ -154,7 +154,33 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
...
+          // template.
+          templateUrl = 'override.html';
+          break;
+        }
+        case 'external': {
+          templateUrl = sourceMapping.templateUrl;
+          break;
+        }
+      }
+    }
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:181]: NIT: call be old fashioned but I would be keen to put this whole block into its own helper function:

```ts
const templateUrl = getTemplateUrl(component);
```

File: packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts [hunk 7]
```diff
@@ -209,7 +210,8 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
       return null;
     }
     return getTemplateMapping(
-        shimSf, positionInShimFile, fileRecord.sourceManager, /*isDiagnosticsRequest*/ false);
+        shimSf, positionInShimFile, fileRecord.sourceManager, /*isDiagnosticsRequest*/ false,
+        this.getTemplateSourceRegistry());
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:240]: OOC why use the `getTemplatSourceRegistry()` in these functions rather than just refer to the private `templateSourceRegistry` directly?

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_source_registry.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * This map keeps track of all template sources that have been type-checked by the id that is
+   * attached to a TCB's function declaration as leading trivia. This enables translation of
+   * diagnostics produced for TCB code to their source location in the template.
+   */
+  private templateSources = new Map<ClassDeclaration, TemplateSource>();
+
+  captureSource(node: ClassDeclaration, mapping: TemplateSourceMapping, file: ParseSourceFile):
+      ClassDeclaration {
+    this.templateSources.set(node, new TemplateSource(mapping, file));
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/template_source_registry.ts:111]: This will still compute and store the line-starts of the `file` for every `node` captured from the file. Perhaps the line-starts should be stored on the `file` itself, to avoid having to do this computation redundantly?

**Review by petebacondarwin — COMMENTED** [review]
Nice refactoring @zarend !

**Review by atscott — CHANGES_REQUESTED** [review]
Just some minor nits. Nice work!

**Comment by zarend** [comment]
FYI: I'm putting this PR on the back-burner for now until we land [40585](https://github.com/angular/angular/pull/40585). That's because @alxhub deleted `overrideComponentTemplate` 💥 , which means much of this PR will need to be rewritten. Ty all for reviewing. I'll re-request reviews when this is ready again.

**Comment by petebacondarwin** [comment]
@zarend - #40585 has now landed. Is it time to restart this PR?
